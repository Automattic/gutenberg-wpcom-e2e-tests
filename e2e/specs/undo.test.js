/**
 * Internal dependencies
 */
import {
	clickBlockAppender,
	getEditedPostContent,
	newPost,
	pressWithModifier,
} from '../support/utils';

describe( 'undo', () => {
	beforeEach( async () => {
		await newPost();
	} );

	it( 'should undo typing after a pause', async () => {
		await clickBlockAppender();

		await page.keyboard.type( 'before pause' );
		await new Promise( ( resolve ) => setTimeout( resolve, 1000 ) );
		await page.keyboard.type( ' after pause' );

		expect( await getEditedPostContent() ).toMatchSnapshot();

		await pressWithModifier( 'primary', 'z' );

		expect( await getEditedPostContent() ).toMatchSnapshot();
	} );

	it( 'should undo typing after non input change', async () => {
		await clickBlockAppender();

		await page.keyboard.type( 'before keyboard ' );
		await pressWithModifier( 'primary', 'b' );
		await page.keyboard.type( 'after keyboard' );

		expect( await getEditedPostContent() ).toMatchSnapshot();

		await pressWithModifier( 'primary', 'z' );

		expect( await getEditedPostContent() ).toMatchSnapshot();
	} );

	it( 'Should undo to expected level intervals', async () => {
		await clickBlockAppender();

		await page.keyboard.type( 'This' );
		await page.keyboard.press( 'Enter' );
		await page.keyboard.type( 'is' );
		await page.keyboard.press( 'Enter' );
		await page.keyboard.type( 'test' );

		expect( await getEditedPostContent() ).toMatchSnapshot();

		await pressWithModifier( 'primary', 'z' ); // Undo 3rd paragraph text.
		await pressWithModifier( 'primary', 'z' ); // Undo 3rd block.
		await pressWithModifier( 'primary', 'z' ); // Undo 2nd paragraph text.
		await pressWithModifier( 'primary', 'z' ); // Undo 2nd block.
		await pressWithModifier( 'primary', 'z' ); // Undo 1st paragraph text.
		await pressWithModifier( 'primary', 'z' ); // Undo 1st block.

		expect( await getEditedPostContent() ).toBe( '' );
		// After undoing every action, there should be no more undo history.
		expect( await page.$( '.editor-history__undo[aria-disabled="true"]' ) ).not.toBeNull();
	} );
} );
