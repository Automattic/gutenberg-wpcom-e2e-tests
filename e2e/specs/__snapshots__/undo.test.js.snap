// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`undo Should undo to expected level intervals 1`] = `
"<!-- wp:paragraph -->
<p>This</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>is</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>test</p>
<!-- /wp:paragraph -->"
`;

exports[`undo should undo typing after a pause 1`] = `
"<!-- wp:paragraph -->
<p>before pause after pause</p>
<!-- /wp:paragraph -->"
`;

exports[`undo should undo typing after a pause 2`] = `
"<!-- wp:paragraph -->
<p>before pause</p>
<!-- /wp:paragraph -->"
`;

exports[`undo should undo typing after non input change 1`] = `
"<!-- wp:paragraph -->
<p>before keyboard <strong>after keyboard</strong></p>
<!-- /wp:paragraph -->"
`;

exports[`undo should undo typing after non input change 2`] = `
"<!-- wp:paragraph -->
<p>before keyboard </p>
<!-- /wp:paragraph -->"
`;
